// bbcode/renderer.ts
// function fixYouTubeLink(link: string) {
//     return "https://www.youtube.com/embed/" + link.match(/([^\/]*$)/)[0];
// }

export function render(node: any): string {
    let result = "";

    node.nodes.forEach(n => {
        if (typeof n === "object") {
            result += render(n)
        } else {
            switch (node.name) {
                case "b": result += bold(n)
            }

            switch (node.name) {
                case "i": result += italic(n)
            }

            switch (node.name) {
                case "u": result += underline(n)
            }

            switch (node.name) {
                case "s": result += strikethrough(n)
            }

            switch (node.name) {
                case "url": result += link(n, node.attrs)
            }
        }
    });

    return result;
}

function bold(value: string): string {
    return `<strong>${value}</strong>`
}

function italic(value: string): string {
    return `<em>${value}</em>`
}

function underline(value: string): string {
    return `<ins>${value}</ins>`
}

function strikethrough(value: string): string {
    return `<del>${value}</del>`
}

function link(value: string, attrs?: any) {
    return `<a href="${attrs.url || value}">${value}</a>`
}
