// api/comment.ts
import { send } from "network";
import type { LikeAction } from "types"

export namespace GetAll {
    export interface Request {
        mandela_id: number;
        limit: number;
        offset: number;
    }

    export interface Comment {
        id: number;
        message: string;
        user_id: number;
        user_name: string;
        create_ts: Date;
        update_ts: Date;
    }

    export interface Response {
        comments: Comment[];
        total_count: number;
    }

    export async function exec(params: Request): Promise<Response> {
        return await send("comment.getAll", params)
    }
}

export namespace Create {
    export interface Request {
        mandela_id: number;
        message: string;
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("comment.create", params)
    }
}

export namespace Update {
    export interface Request {
        id: number;
        message: string;
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("comment.update", params)
    }
}

export namespace Delete {
    export interface Request {
        id: number;
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("comment.delete", params)
    }
}

export namespace Like {
    export interface Request {
        id: number;
        action: LikeAction
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("comment.like", params)
    }
}
