// api/user.ts
import { send } from "network";

export namespace GetOne {
    export interface Response {
        code: string;
        create_ts: Date;
        id: number;
        name: string;
    }

    export async function exec(): Promise<Response> {
        return await send("user.getOne");
    }
}

export namespace Create {
    export interface Request {
        id: number,
        name: string,
        code: string,
        token: string,
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("user.create", params);
    }
}

export namespace Update {
    export interface Request {
        code: string;
        name: string;
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("user.update", params);
    }
}

export namespace UpdateToken {
    export interface Request {
        token: string;
    }

    export async function exec(params: Request): Promise<{}> {
        return await send("user.updateToken", params);
    }
}

export namespace GetNextId {
    export interface Response {
        id: number;
    }

    export async function exec(): Promise<Response> {
        return await send("user.getNextId")
    }
}

export namespace Auth {
    export interface Request {
        token: string;
    }

    export interface Response {
        code: string,
        name: string,
    }

    export async function exec(params: Request): Promise<Response> {
        return await send("user.auth", params);
    }
}

export namespace Logout {
    export async function exec(): Promise<{}> {
        return await send("user.logout");
    }
}
